src/node/stage/approved.rs

```diff
@@ -781,6 +840,67 @@ impl Approved {
         }
     }
 
+    // TODO: accumulate at least quorum of these messages
+    //       and only then proceed to handle the DKG result.
+    pub fn handle_dkg_old_elders(
+        &mut self,
+        core: &Core,
+        participants: BTreeSet<PublicId>,
+        parsec_version: u64,
+        public_key_set: bls::PublicKeySet,
+        _src_id: PublicId,
+    ) -> Result<()> {
+        debug!(
+            "notified by DKG participants {:?} to vote for SectionInfo",
+            participants
+        );
+
+        let dkg_result = DkgResult::new(public_key_set, None);
+        if self
+            .dkg_voter
+            .has_info(&(participants.clone(), parsec_version), &dkg_result)
+        {
+            self.handle_dkg_result_event(core, &participants, &dkg_result)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn handle_dkg_message(
+        &mut self,
+        core: &mut Core,
+        participants: BTreeSet<PublicId>,
+        parsec_version: u64,
+        message_bytes: Bytes,
+        pub_id: PublicId,
+    ) -> Result<()> {
+        trace!(
+            "handle dkg message of p{:?}-{:?} from {}",
+            participants,
+            parsec_version,
+            pub_id
+        );
+        self.dkg_voter
+            .set_timer_token(core.timer.schedule(DKG_PROGRESS_INTERVAL));
```

Comment:
I wonder, could a malicious node keep sending invalid DKG messages to us to prevent the DKG from ever expiring? Shouldn't we somehow verify the message is valid before extending the timer? Not sure if that makes sense. Or alternatively, don't extend the timer here, only in `handle_timeout`, so `progress_dkg` is always called?

Reply:
an incoming DkgMessage could case key_generator to transit into next phase as well, which means the ticker shall be set to a new token.
but your concern is valid.
I will move it into the place that only a valid DkgMessage will reset the ticker.